src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -424,34 +425,38 @@ getFloatingRate s calculationPeriod paymentPeriod
         rateFixingCalendar
         s.calculationPeriodDates.calculationPeriodDatesAdjustments.businessDayConvention
         fixLeg
+        rateFixingDate
     _ -> regularRate
 
 -- | HIDE
 -- Recursive function to calculate flat compounding.
 -- It follows the ISDA calculation example for flat compounding, example 4.2:
 -- https://www.isda.org/a/2KiDE/isda-compounding-memo.pdf
 calcFlatCompounding : O -> Deliverable -> SchedulePeriod -> SwapStream
-  -> O -> O -> HolidayCalendarData -> Bool ->
+  -> Text -> Decimal -> HolidayCalendarData -> Bool ->
   [CalculationPeriodData] -> [CalculationPeriodData] -> [C]
 calcFlatCompounding notional currency p s
-  regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+  floatingRateIndex floatingRateSpread rateFixingCalendar fixLeg
   (x::xs) calculationPeriodDataSelected =
   calcFlatCompounding notional currency p s
-    regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+    floatingRateIndex floatingRateSpread rateFixingCalendar fixLeg
     xs calculationPeriodDataSelected ++
   calcFlatCompounding notional currency p s
-  regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+  floatingRateIndex floatingRateSpread rateFixingCalendar fixLeg
   xs (calculationPeriodDataSelected ++ [x])
 calcFlatCompounding notional currency p s
-  regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
-  [] calculationPeriodDataSelected  =
+  floatingRateIndex floatingRateSpread rateFixingCalendar fixLeg
+  [] calculationPeriodDataSelected =
   if null calculationPeriodDataSelected then [] else [scale notional $ compoundedRates]
     where
     foldRates acc (pp, includeSpread) =
       when (TimeGte pp.rateFixingDate) $ scale (rate * pp.dcf) acc
```

Comment:
same here, we can probable refactor this?

Reply:
interesting. Could you push to a branch for me to have a look at?

Reply:
The commit [remove when node from calcFlatCompounding](https://github.com/digital-asset/daml-finance/pull/812/commits/9971e8e5fd0d2c0a31374d80d99df8362f6ee21a) below should give the error, at least it does for me locally.

Reply:
If we remove the `when` nodes that fix the observation date, we should then move the resulting `scale` node after the `when` node used for the payment. 

Specifically, from
- when fixing $ scale obs $ when payment $ one usd
to
- when payment $ scale obs $ one usd

Reply:
Now it works, thanks!